### PR TITLE
Add support for SpeakerPosition with Cartesian coordinates

### LINKED ISSUE #20: AudioBlockFormatDirectSpeakers only supports polar SpeakerPositions
The ability for `DirectSpeaker`s `audioBlockFormat`s to contain Cartesian positions is mentioned at the end of BS.2076-1 section 5.4.3.1.

I don't desperately need this, but turning `SpeakerPosition` into a variant will be an API change, so should ideally happen before 1.0.

### PR DESCRIPTION
**Pull request by rsjtaylor**
Details:

1) Renamed SpeakerPosition SphericalSpeakerPosition
2) Added CartesianSpeakerPosition with appropriate coordinate types
3) Made SpeakerPosition a variant of the two types of speaker position
4) Updated AudioBlockFormatDirectSpeakers and the xml parse/format methods to use the new types.
5) Updated existing tests to use SphericalSpeakerPosition (no extra testing for CartesianSpeakerPosition atm)

The parsing is slightly messy as the existing methods don't handle a variant type and you can only figure out which type it should be from the attribute values during parsing (there's no Cartesian flag like there is in AudioBlockFormatObjects), which breaks the existing validator system
I've currently got get/set/has for both the concrete position types and the variant as I figured they're useful in different contexts, but maybe I should only have the position types in the API and keep the variant as an implementation detail?
It's not clear (to me) from the spec whether not specifying azimuth is illegal when Cartesian coordinates are present, or whether having both types of coordinates is OK - I've assumed you don't need azimuth and you can't have both here.
Could potentially remove almost all of the duplication by parameterising the types/parsing on the coordinate system, but this would be a bit inconsistent with the rest of the library and maybe overcomplicating things?

Closes #20

### CODE REVIEW

File: include/adm/elements/speaker_position.hpp [hunk 1]
```diff
@@ -20,16 +21,177 @@ namespace adm {
...
+     * argument. Removes an ADM parameter if it is optional or resets it to
+     * the default value if there is one.
+     */
+    template <typename Parameter>
+    void unset();
+
+    /**
+     * @brief Print overview to ostream
+     */
+    void print(std::ostream &os) const;
```
Comment (tomjnixon) [inline include/adm/elements/speaker_position.hpp:104]: I don't think this is implemented? Just removing it is fine by me.

**Review by tomjnixon — COMMENTED** [review]
Thanks, I think this looks good overall!

> I've currently got get/set/has for both the concrete position types and the variant as I figured they're useful in different contexts, but maybe I should only have the position types in the API and keep the variant as an implementation detail?

This sounds fine to me; it should make it map to libear a bit more nicely.

> It's not clear (to me) from the spec whether not specifying azimuth is illegal when Cartesian coordinates are present, or whether having both types of coordinates is OK - I've assumed you don't need azimuth and you can't have both here.

I'm pretty sure it should be one or the other. There are examples with only Cartesian coordinates (so, azimuth is not required), and the renderer doesn't have any concept of the output being cartesian or polar, so having both doesn't make sense there either.

On this topic, should `speakerPosition_` be a `boost::optional`? It seems like it should just be one or the other. I suppose that makes the semantics of `unset` different, but IMO that doesn't really make sense currently -- why would you want to unset the speakerPosition except if you want to change it to the other type?

Just having get(), set(), and has() for this would be my preference.

I wasn't sure what to make of isDefault for this as well -- is it required that this is implemented for all fields?

> Could potentially remove almost all of the duplication by parameterising the types/parsing on the coordinate system, but this would be a bit inconsistent with the rest of the library and maybe overcomplicating things?

That sounds like it may end up significantly more complicated, which doesn't seem worth it to me.

Some other things I spotted that are worth fixing:

-   If I'm reading it right, the formatter assumes that the xml looks like:

    ```xml
    <position coordinate="azimuth" min="15.0" max="45.0">30.0</position>
    ```

    whereas the parser (and the EAR) assumes that it's:

    ```xml
    <position coordinate="azimuth">30.0</position>
    <position coordinate="azimuth" bound="min">15.0</position>
    <position coordinate="azimuth" bound="max">45.0</position>
    ```

    I think this should be clarified; the only clues that I can find are that the old bbcat libraries had an example of doing it the second way, and there's other bound attributes in 2076 which are specified the same, but have examples showing the second way.

- I'd prefer to merge #57 for the algorithm include fix instead, as in theory this is needed in lots of files.

- It would be nice to have a test for the parser, including the bounds. Testing the formatter would be good too, but that's possibly a bit trickier.

... and a couple of minor nits that I'm happy to ignore:

- It should probably be squashed into a single commit?

- It might be worth running through git-clang-format? This seems to do a good job of only fixing changed lines. I can do this before merging if it's a pain.

File: tests/test_data/write_specified_cartesian_speaker.accepted.xml [hunk 2]
```diff
@@ -0,0 +1,24 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (rsjtaylor) [inline tests/test_data/write_specified_cartesian_speaker.accepted.xml:1]: @tomjnixon does this file now show the output you would expect? Particularly the bound="min" etc.. bits and the screenEdgeLock?

File: tests/test_data/write_specified_cartesian_speaker.accepted.xml [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+						<position coordinate="Z">0.500000</position>
+						<position coordinate="Z" bound="min">0.400000</position>
+						<position coordinate="Z" bound="max">0.600000</position>
+					</audioBlockFormat>
+				</audioChannelFormat>
+			</audioFormatExtended>
+		</format>
+	</coreMetadata>
+</ebuCoreMain>
+
```
Comment (rsjtaylor) [inline tests/test_data/write_specified_cartesian_speaker.accepted.xml]: Helps if I highlight the whole thing.

Comment (tomjnixon) [reply tests/test_data/write_specified_cartesian_speaker.accepted.xml]: yeah that looks good to me (or at least it matches the test cases in the EAR)

File: src/private/rapidxml_formatter.cpp [hunk 4]
```diff
@@ -299,36 +305,122 @@ namespace adm {
...
+            node.setValue(max.get());
+        });
+    }
+
+    void formatCartesianSpeakerPosition(
+        XmlNode &parentNode, const std::string &name,
+        const CartesianSpeakerPosition &position) {
+
+      auto xNode = parentNode.addNode(name);
+        xNode.addAttribute("coordinate", "X");
```
Comment (tomjnixon) [inline src/private/rapidxml_formatter.cpp:371]: Could wrap these blocks with `{}` if you want this indent?

File: src/private/xml_parser.cpp [hunk 5]
```diff
@@ -417,12 +418,95 @@ namespace adm {
...
+          if(axis == std::string("X")
+              || axis == std::string("Y")
+              || axis == std::string("Z")) {
+            cartesianCoordinates.emplace_back(element, axis);
+          }
+          if(axis == std::string{"azimuth"}
+              || axis == std::string{"elevation"}
+              || axis == std::string{"distance"}) {
+            sphericalCoordinates.emplace_back(element, axis);
+          }
```
Comment (tomjnixon) [inline src/private/xml_parser.cpp:441]: perhaps add error condition if the axis doesn't match any of these?

File: src/private/xml_parser.cpp [hunk 6]
```diff
@@ -417,12 +418,95 @@ namespace adm {
...
+        if (axe == "X") {
+          if (bound == boost::none) {
+            adm::xml::setValue<adm::X>(element, speakerPosition);
+            adm::xml::setOptionalAttribute<adm::HorizontalEdge>(element, "screenEdgeLock",
+                                                                screenEdgeLock);
+          } else if (bound.get() == "min") {
+            adm::xml::setValue<adm::XMin>(element, speakerPosition);
+          } else if (bound.get() == "max") {
+            adm::xml::setValue<adm::XMax>(element, speakerPosition);
+          }
```
Comment (tomjnixon) [inline src/private/xml_parser.cpp:480]: worth raising an error if the bound is neither min or max?

**Comment by tomjnixon** [comment]
Hey, this looks good -- thanks for making the changes.

I don't mind too much about the checks I suggested, it looks like they aren't done for polar either.

Squash it and run it through clang-format-diff, and I'm happy.

File: src/private/rapidxml_formatter.cpp (hunk 4 above)
Comment (rsjtaylor) [reply src/private/rapidxml_formatter.cpp:371]: Thanks, just a formatting error I think (Have discovered I was a bit confused about how my IDE applies auto-indents) so will let clang-format sort that one out.

File: src/private/xml_parser.cpp (hunk 6 above)
Comment (rsjtaylor) [reply src/private/xml_parser.cpp:493]: Turns out that already happened via the validator on the Bound class. Have changed things slightly by catching the `InvalidString` the validator throws and throwing an `XmlParsingError` instead as it's slightly more helpful & reports the line number. Also added a test.

**Pull request merged**